code/game/machinery/doppler_array.dm

```diff
@@ -88,4 +154,54 @@ var/list/doppler_arrays = list()
 			stat &= ~NOPOWER
 		else
 			icon_state = "[initial(icon_state)]-off"
-			stat |= NOPOWER
\ No newline at end of file
+			stat |= NOPOWER
+
+/obj/machinery/doppler_array/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "doppler_array.tmpl", "Tachyon-doppler array", 500, 650)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/machinery/doppler_array/ui_data(mob/user, ui_key = "main", datum/topic_state/state = default_state)
+	var/data[0]
+	var/list/explosion_data = list()
+	for(var/datum/explosion_log/E in logged_explosions)
+		E.UID()

```

Comment:
@Crazylemon64 I have moved `UID()` up to here

Reply:
@Citinited That isn't how UID works.... `unique_datum_id` shouldn't ever be accessed directly. UID() returns the id for you. You can just replace `"unique_datum_id" = E.unique_datum_id` with `"unique_datum_id" = E.UID()`

Reply:
Aah, right. I'll change that.

Out of interest, why shouldn't you directly access the `unique_datum_id`?

Reply:
It's a meta var, it's meant purely for the UID system to interact with. It'd be better off being named something like __id, tbh

Reply:
Moved!